### PR TITLE
Allow ability to override max Metric streams and MetricPoints per stream

### LINKED ISSUE #2358: Allow configuring max Metrics allowed, max MetricPoints allowed per Metric
These are currently hardcoded as 1000, 2000 respective. The behavior after the limits are hit are dropping measurements on the floor.

This issue is to track the following:
1. Decide if and how configuring this should be exposed to users.
2. Decide a good default

Might have spec dependency.

### PR DESCRIPTION
**Pull request by cijothomas**
Fixes #2358 
This allows user to specify a max limit of number of Metric streams. This is to prevent indefinite memory growth, if a user keeps creating new instruments, instead of reusing for each measurement.

This also allows user to specify a max limit of the number of MetricPoints, *per* metric stream. This is to prevent indefinite memory usage growth, when high cardinality values are uses in metric tags. This limit is applied independently to every metric stream for this meter provider. In the future, we could add ability to configure separate MaxMetricPoint limit *per* instrument, by leveraging Views, which can override the `MeterProvider` level limits.

Note that, in the current implementation, MetricPoints are pre-allocated upto max limit when an instrument is created. So user pays upfront memory cost for MaxMetricPoints even though only subset of them is used. In the future, when per-instrument config is allowed, users can tune this number to reduce wastage.

Also as [this](https://github.com/open-telemetry/opentelemetry-dotnet/issues/2360), is not addressed, there is no ability to "reclaim" a MetricPoint if it has been inactive for some time.

TODO:
Public API
Unit Tests
Decide if the defaults 1000, 2000 are good enough.

### CODE REVIEW

File: src/OpenTelemetry/Metrics/AggregatorStore.cs
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs:59]: I think metricPoinLimit+1 should be used, as we reserve one point for zero dimension case.

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (alanwest) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:149]: Bit of a nit, but this is the first time we're publicly introducing the term MetricPoint. So maybe

```suggestion
        /// Sets the maximum number of metric points allowed per metric stream.
```

I think the name of the method is fine `SetMetricPointPerMetricStreamLimit` ... unless there's a better term from the spec for describing the `MetricPoint` concept?

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (alanwest) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:151]:
```suggestion
        /// This limits the number of unique combinations of key/value pairs used
        /// when reporting measurements.
```

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (alanwest) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:159]: Not sure I understand this remark.

**Review by alanwest — APPROVED** [review]
No reason to hold up this PR, but I'm wondering if we should change the name `MetricPoint` to something else?


The term "metric stream" seems pretty well established in the specification, but "metric point" is not as well established. In fact there is a section of the specification here titled [`MetricPoint`](https://github.com/open-telemetry/opentelemetry-specification/blob/1e53ca9fb79e5a3d682c550d69c0a2c1dfb553ca/specification/metrics/datamodel.md#metric-points), but it refers to a different concept than what we've modeled with the MetricPoint class.

We had used the term `DataPoint` in the past. Searching that data model specification "data point" is a term that is often used. So, what do folks think about maybe `DataPoint` or `MetricDataPoint`?

File: src/OpenTelemetry/Metrics/AggregatorStore.cs
Comment (utpilla) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:59]: Would it not be misleading for the user if we add +1 here? Maybe we could update the remarks for the public method `SetMetricPointPerMetricStreamLimit` in that case?

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (reyang) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:149]: Personal gut feeling - maybe for the user "cardinality" is something easier to digest?

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (reyang) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:157]:
```suggestion
        /// If a particular combination of key value pair is used at least once,
```

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (utpilla) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:149]: How about `SetTimeSeriesLimit`?

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (utpilla) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:127]:
```suggestion
        /// so this controls the number of instruments supported.
```

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (utpilla) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:128]:
```suggestion
        /// When Views are configued, a single instrument can result in multiple metric streams,
```

File: src/OpenTelemetry/Metrics/AggregatorStore.cs
Comment (reyang) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:59]: > I think metricPoinLimit+1 should be used, as we reserve one point for zero dimension case.

That doesn't seem to be mathematically correct?

Comment (utpilla) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:59]: We would throw if the provided value is less than 1 as the method validates the input:
`Guard.Range(metricPointLimit, min: 1);`

I think what could be confusing is that the user provides a value thinking that it will be maximum number of time-series allowed but internally we would be allowing an additional time-series.

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:159]: Current behaviour is - if you ever use a KVP, then its counted against the cap. Once https://github.com/open-telemetry/opentelemetry-dotnet/issues/2360 is addressed, we'll be able to totally forgot unused ones, and then it'd no longer contribute to the cap.

File: src/OpenTelemetry/Metrics/MeterProviderBuilderBase.cs
Comment (CodeBlanch) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderBase.cs:112]: Just a random thought, we could combine these two methods into a single one something like:

```csharp
SetMetricLimits(int metricStreamLimit = [default], int metricPointLimit = [default])
```

Kind of reminds me of: https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadpool.setmaxthreads?view=net-6.0

As far as the names, I think "min/max" might be slightly more standard than "limit"?

`maxUniqueMetricStreamCount`
`maxUniqueMetricPointCount`

Or maybe "capacity"?

`uniqueMetricStreamCapacity`
`uniqueMetricPointCapacity`

Capacity might better indicate it is pre-allocated. Common in the stream apis ex: https://docs.microsoft.com/en-us/dotnet/api/system.io.memorystream.-ctor?view=net-6.0#System_IO_MemoryStream__ctor_System_Int32_

🤷

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderBase.cs:112]: renaming to SetMaxMetricStreams and SetMaxMetricPointsPerMetricStream as they sound better.
Capacity might indicate the pre-allocation, but that's a implementation detail and may change.

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:149]: maxTimeseries seems a good option.

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:149]: cardinality - i think this would be a better fit, if the limit was for limiting the cardinality of an individual tag.

File: src/OpenTelemetry/Metrics/MeterProviderBuilderBase.cs
Comment (reyang) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderBase.cs:32]: Remove this variable since the name is misleading (it is not the max, it's just the default of the limit).

File: src/OpenTelemetry/Metrics/MeterProviderBuilderBase.cs
Comment (reyang) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderBase.cs:112]: Is there a reason to use different names other than `SetMetricStreamsLimit(int limit)` (keeping the method name and the parameter name aligned)?

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderBase.cs:112]: no. just typo. will fix.

File: src/OpenTelemetry/Metrics/AggregatorStore.cs
Comment (alanwest) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:59]: Should we consider doing away with the reservation of the zero dimension case? I guess I don't know the history of why we do it.

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (alanwest) [inline src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:159]:
```suggestion
        /// If a particular key/value pair combination is used at least once,
        /// it will contribute to the limit for the life of the process.
        /// This may change in the future. See: https://github.com/open-telemetry/opentelemetry-dotnet/issues/2360.
```

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (alanwest) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs]: Gotcha, figured this is what you meant. What do you think about this clarification https://github.com/open-telemetry/opentelemetry-dotnet/pull/2635#discussion_r752541574

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (reyang) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:149]: Check this out https://en.wikipedia.org/wiki/Cardinality

**Review by reyang — APPROVED** [review]

File: src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/MeterProviderBuilderExtensions.cs:149]: lets me proceed as is now, to unblock release.
Will discuss the naming issues in next SIG meeting. We have ~10 more days before planned stable release, for coming up with better names)

**Pull request merged**
